python/tvm/testing.py

```diff
@@ -285,4 +288,184 @@ def _check_forward(constraints1, constraints2, varmap, backvarmap):
                    constraints_trans.dst_to_src, constraints_trans.src_to_dst)
 
 
+def uses_gpu(f):
+    """Mark to differentiate tests that use the GPU is some capacity.
+
+    These tests will be run on CPU-only test nodes and on test nodes with GPUS.
+    To mark a test that must have a GPU present to run, use `@requires_gpu`.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.gpu(f)
+
+
+def requires_gpu(f):
+    """Mark a test as requiring a GPU to run.
+
+    Tests with this mark will not be run unless a gpu is present.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.skipif(not tvm.gpu().exist, reason="No GPU present")(uses_gpu(f))
+
+
+def requires_cuda(f):
+    """Mark a test as requiring the CUDA runtime.
+
+    This also marks the test as requiring a gpu.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.cuda(
+        pytest.mark.skipif(
+            not tvm.runtime.enabled("cuda"), reason="CUDA support not enabled"
+        )(requires_gpu(f))
+    )
+
+
+def requires_opencl(f):
+    """Mark a test as requiring the OpenCL runtime.
+
+    This also marks the test as requiring a gpu.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.opencl(
+        pytest.mark.skipif(
+            not tvm.runtime.enabled("opencl"), reason="OpenCL support not enabled"
+        )(requires_gpu(f))
+    )
+
+
+def requires_rocm(f):
+    """Mark a test as requiring the rocm runtime.
+
+    This also marks the test as requiring a gpu.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.rocm(
+        pytest.mark.skipif(
+            not tvm.runtime.enabled("rocm"), reason="rocm support not enabled"
+        )(requires_gpu(f))
+    )
+
+
+def requires_tensorcore(f):
+    """Mark a test as requiring a tensorcore to run.
+
+    Tests with this mark will not be run unless a tensorcore is present.
+
+    Parameters
+    ----------
+    f : function
+        Function to mark
+    """
+    return pytest.mark.tensorcore(
+        pytest.mark.skipif(
+            not tvm.gpu().exist or not nvcc.have_tensorcore(tvm.gpu(0).compute_version),
+            reason="No tensorcore present",
+        )(f)
+    )
+
+
+def parametrize_devices(f):
+    """Parametrize a test over all enabled devices.
+
+    Parameters
+    ----------
+    f : function
+        Function to parametrize. Must be of the form `def test_xxxxxxxxx(device, ctx)`:,
+        where `xxxxxxxxx` is any name.
+
+    Example
+    -------
+    >>> @tvm.testing.parametrize
+    >>> def test_mytest(device, ctx):
+    >>>     ...  # do something
+    """
+    return pytest.mark.parametrize("device,ctx", enabled_devices())(uses_gpu(f))
```

Comment:
Remove the `uses_gpu`.

Reply:
`uses_gpu` is necessary here. Any function using `enabled_devices` must use `uses_gpu` otherwise it will not be run on the gpu nodes. I've tried to provide an explanation here: https://discuss.tvm.ai/t/rfc-testing-split-testing-based-on-cpu-gpu/7722.